### PR TITLE
Allow data driven styles on circles

### LINKED ISSUE #1321: Enable data driven styling for `circle` layers

### PR DESCRIPTION
**Pull request by lucaswoj**
Resolves #1321 

### Remaining Tasks

 - [x] Fix disappearing / wrong colors 
 - [x] Fix [antialiasing on heterogeneously sized circles](https://cloud.githubusercontent.com/assets/281306/8336674/fdd21f5a-1a5a-11e5-9caf-f6d4f5faaa96.png)

### CODE REVIEW

**Comment by jfirebaugh** [comment]
CC @ansis for review.

**Comment by lucaswoj** [comment]
Ready for :eyes: @jfirebaugh @ansis 

There are still some problems running against `gl-test-suite` but they're either out of scope for this PR or need to be fixed upstream. 

Note that I refactored the `util/browser` module to allow it to function somewhat in the context of a Web Worker. And I refactored some color parsing logic that was causing bugs and a little hard to understand.

**Comment by lucaswoj** [comment]
screen shot 2015-06-24 at 2 23 06 pm
9780375980176_p0_v1_s260x420

**Comment by jfirebaugh** [comment]
The stuff in `browser` pretty much isn't going to work in a worker context -- I'd prefer not requiring `browser` altogether to faking `browser.devicePixelRatio`. Is that feasible?

**Comment by lucaswoj** [comment]
@jfirebaugh How would you go about faking `browser.devicePixelRatio`? My first instinct was to pass it down to `WorkerTile` but that involved a lot of not-so-semantic code changes. Is there a good place for global state within a worker?

**Comment by jfirebaugh** [comment]
Is it possible to apply the `devicePixelRatio` at render time rather than layout time?

**Comment by lucaswoj** [comment]
Unfortunately not. We need the `devicePixelRatio` to correctly populate the vertex buffer.

**Comment by lucaswoj** [comment]
We could potentially

 - ignore the `devicePixelRatio` altogether 
 - calculate the antialiasing within the shader itself (?)
 - do antialiasing another way altogether (?)
 - populate the vertex buffer at render time

**Comment by ansis** [comment]
for devicePixelRatio I think we should either
- pass it to WorkerTile and then on to either CircleBucket or CircleVertexBuffer, depending on where you want to do the adding
- create a uniform, pass the value to the vertex shader and add it to the buffer value there

File: js/render/draw_circle.js [hunk 1]
```diff
@@ -4,6 +4,30 @@ var browser = require('../util/browser.js');
...
+    },
+    {
+        styleName: 'circle-radius',
+        styleType: 'number',
+        glName: 'a_size',
+        glWidth: 2,
+        glType: '1f'
+    }
+];
+
```
Comment (ansis) [inline js/render/draw_circle.js]: I have mixed feelings about this. I sort of like it but I also feel that it adds another level of abstraction without making things clearer.

Comment (lucaswoj) [inline js/render/draw_circle.js]: This was a small trial of a larger refactoring I want to do of the buffer generation code. Building better abstractions for buffer generation will make our code shorter and therefore less buggy and easier to maintain. Tracking the larger issue in #1318 (though there aren't any more details there right now).

Comment (ansis) [inline js/render/draw_circle.js]: Yep. I'm looking forward to seeing all of it!

**Comment by lucaswoj** [comment]
`84 failed, 166 passed assertions.` :ship: :shipit: :+1:

**Comment by lucaswoj** [comment]
@ansis @jfirebaugh I removed the `browser` tomfoolery in https://github.com/mapbox/mapbox-gl-js/commit/29d99d0f6b2a3a25ba7b8c66b73e5324fe0f7a87. Is this ready to go?

**Comment by ansis** [comment]
Yep, looks good to me!

**Pull request merged**